src/main/scala/com/softwaremill/realworld/users/UsersService.scala

```diff
@@ -26,24 +25,35 @@ class UsersService(usersRepository: UsersRepository):
     def checkUserDoesNotExist(email: String): IO[Exception, Unit] =
       for {
         maybeUser <- usersRepository.findByEmail(email.toLowerCase)
-        _ <- ZIO.fail(Exceptions.Conflict(EmailAlreadyUsed)).when(maybeUser.isDefined)
+        _ <- ZIO.fail(Exceptions.AlreadyInUse("E-mail already in use!")).when(maybeUser.isDefined)
       } yield ()
 
     def generateJwt(): ZIO[Any, Nothing, String] = ZIO.succeed("tempJWT") // TODO add JWT generation
-    def encryptPassword(password: String): ZIO[Any, Nothing, String] = ZIO.succeed(password) // TODO add encryption
 
     for {
       _ <- checkUserDoesNotExist(emailClean)
       user <- {
         for {
-          bcryptPassword <- encryptPassword(passwordClean)
+          hashedPassword <- PasswordHashing.encryptPassword(passwordClean)
           jwt <- generateJwt()
-          _ <- usersRepository.add(UserRegisterData(emailClean, usernameClean, bcryptPassword))
+          _ <- usersRepository.addUser(UserRegisterData(emailClean, usernameClean, hashedPassword))
         } yield User(userWithToken(emailClean, usernameClean, jwt))
       }
     } yield user
   }
 
+  def userLogin(user: UserLoginData): IO[Exception, UserData] = {
+    val emailClean = user.email.toLowerCase.trim()
+    val passwordClean = user.password.trim()
+
+    for {
+      maybeUser <- usersRepository.findUserWithPasswordByEmail(emailClean)
+      _ <- ZIO.fail(Exceptions.InvalidCredentials()).when(maybeUser.isEmpty)
+      userWithPassword = maybeUser.get // TODO should be changed
```

Comment:
indeed, should be :) You can e.g. do `ZIO.fromOption` + `.mapError`

Reply:
Thanks, fixed :)